Add additional kwargs support for Cohere SDK

giannis2two opened the pull request:
* Adds support for `additional_kwargs` in `get_cohere_chat_request`
  * This functionality passes in Cohere SDK specific parameters to the API request
* Added notebook for `connectors` to demonstrate the usage of `additional_kwargs` params

ccurme committed:
add kwargs (#19519)

`HanaDB.add_texts` is missing **kwargs.

znwilkins committed:
langchain: Passthrough batch_size on index()/aindex() calls (#19443)

**Description:** This change passes through `batch_size` to
`add_documents()`/`aadd_documents()` on calls to `index()` and
`aindex()` such that the documents are processed in the expected batch
size.
**Issue:** #19415
**Dependencies:** N/A
**Twitter handle:** N/A

giannis2two committed:
Add additional kwargs support for Cohere SDK

giannis2two committed:
Add connectors notebook

giannis2two committed:
update notebook

harry-cohere commented on libs/partners/cohere/langchain_cohere/chat_models.py:
```diff
@@ -60,19 +58,17 @@ def get_cohere_chat_request(
+        }
+        for i, doc in enumerate(additional_kwargs.get("documents", []) or [])
+    ]
+    connectors = [{"id": connector} for connector in additional_kwargs.get("connectors", []) or []]
```
`None` is now filtered out from being sent to the API
```suggestion
    connectors = [{"id": connector} for connector in additional_kwargs.get("connectors", []) or None]
```

harry-cohere replied:
actually, sorry, the `if not connectors: connectors = None` is a slightly better approach than the suggestion I made

giannis2two replied:
👍

harry-cohere commented on libs/partners/cohere/langchain_cohere/chat_models.py:
```diff
@@ -60,19 +58,17 @@ def get_cohere_chat_request(
+            "id": doc.metadata.get("id") or f"doc-{str(i)}",
+        }
+        for i, doc in enumerate(additional_kwargs.get("documents", []) or [])
+    ]
```
```suggestion
    ]
    if not documents:
        documents = None
```

harry-cohere commented on libs/partners/cohere/langchain_cohere/chat_models.py:
```diff
@@ -60,19 +58,17 @@ def get_cohere_chat_request(
+    additional_kwargs = messages[-1].additional_kwargs
+
+    # cohere SDK will fail loudly if both connectors and documents are provided
+    documents = [
```
What are your thoughts on adding `documents` and `connectors` to `ChatCohere` ?

giannis2two replied:
Hmm I don't think that would be needed. The sense I have been getting from the LC people is that they like having such inference based parameters (documents, connectors) on a "per message" basis vs generic ChatModel configurations. That way with new instantiations of eg. `HumanMessage` you can configure which parameters you want to include. So that means that in a conversation history you can have multiple turns where say first turn used documents, second turn used connectors and third turn didn't use anything.

harry-cohere replied:
That makes sense!

giannis2two committed:
nits

eyurtsev committed:
core[patch]: Use defusedxml in XMLOutputParser (#19526)

This mitigates a security concern for users still using older versions of libexpat that causes an attacker to compromise the availability of the system if an attacker manages to surface malicious payload to this XMLParser.

billytrend-cohere committed:
cohere[patch]: add cohere as a partner package (#19049)

Description: adds support for langchain_cohere

---------

Co-authored-by: Harry M <127103098+harry-cohere@users.noreply.github.com>
Co-authored-by: Erick Friis <erick@langchain.dev>

sepiatone committed:
core[runnables]: docstring for class runnable, method with_listeners() (#19515)

**Description:** Docstring for method with_listerners() of class
Runnable
**Issue:** [Add in code documentation to core Runnable methods
#18804](https://github.com/langchain-ai/langchain/issues/18804)
**Dependencies:** None

Commit:
cohere[patch]: fix release (#19529)

giannis2two committed:
notebook nit

giannis2two committed:
Merge branch 'master' of https://github.com/langchain-ai/langchain into giannis/additional_kwargs
